### DISCUSSION TITLE
Help with getting started with Faust and STK

### DISCUSSION THREAD

**Discussion by andersmattson** [pschatzmann/arduino-audio-tools#697] [Q&A]
I'm trying to get any kind of reverb working on a Esp32 Wrover-E and I can't really get my head around what I'm doing wrong. 
I'm trying to put this effect between two I2S -> I2S streams (code below).

First I tried with the STK effects that comes with the repo and after some compilation errors I got the STKNReverb to compile (compiler complained about the class being abstract but it worked after adding the clone function to the class in the source code, maybe a bug?). Running that outputs an error several times and then no audio output:
```
"Error:   resize: Not enough memory for 4096 frames"
```
which made me think that the STK effects maybe isn't optimized for the Wrover's PSRAM as I understand the Faust implementation is.

So I tried some simple Faust programs but the only one I could get to work was a simple echo:
```
import("stdfaust.lib");
process = ef.echo1s;
```

When trying with something a bit more elaborate I only get silence. I did get the Guitarix example DSP sort of working but the audio was really bad when applied to any input sound (sine or square wave coming in via I2S).

So, to make it brief, what does it take to get for instance the following DSP code to work on a WROVER? It works perfectly fine in the Faust Online IDE).
```
import("stdfaust.lib");
process = re.stereo_freeverb(0.8, 0.9, 0.6, 100);
```

Here's my code:

```
#define DEFAULT_BUFFER_SIZE 128
#define I2S_BUFFER_SIZE 64
#define I2S_BUFFER_COUNT 8

#define USE_MEMORY_MANAGER

#include "AudioTools.h"
#include "AudioLibs/AudioFaust.h"

#include "mydsp.h"

typedef int16_t sound_t;
uint16_t sample_rate=44100;
uint8_t channels = 2;

I2SStream in;
I2SStream out;
FaustStream<mydsp> faust(out);

StreamCopy copier(faust, in);

void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);


    auto fcfg = faust.defaultConfig();
    faust.begin(fcfg);

    auto cfg = in.defaultConfig(RX_MODE);
    cfg.i2s_format = I2S_STD_FORMAT;
    cfg.port_no = 0;
    cfg.bits_per_sample = 16;
    cfg.channels = channels;
    cfg.sample_rate = sample_rate;
    cfg.pin_bck = 27;
    cfg.pin_ws = 25;
    cfg.pin_data = 26;
    cfg.is_master = false;
    cfg.use_apll = false;
    in.begin(cfg);

    auto config_out = out.defaultConfig(TX_MODE);
    config_out.sample_rate = sample_rate; 
    config_out.bits_per_sample = 16;
    config_out.i2s_format = I2S_STD_FORMAT;
    config_out.is_master = true;
    config_out.port_no = 1;
    config_out.pin_bck = 18;
    config_out.pin_ws = 19;
    config_out.pin_data = 23;
    out.begin(config_out);
}

void loop() {
    copier.copy();
}
```

**Comment by pschatzmann** [comment]
Yes, there is no special logic for PSRAM in STK but you can try to use [this trick, that might work](https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/).

I am not sure what the issue with the Faust example is...
